xsd/pointwise_models/zscore.py

```diff
@@ -0,0 +1,226 @@
+            Training historical model data
+        y : pd.Series or pd.DataFrame, shape (n_samples, 1)
+            Target measured values.
+        window_width :  The size of the moving window for 
+            statistical analysis. Default is 31 days.
+
+        Returns
+        -------
+        self : returns an instance of self.
+        """
+
+        X_mean, X_std = _calc_stats(X, self.window_width)
+        y_mean, y_std  = _calc_stats(y, self.window_width)
+        
+        shift, scale = _get_params(X_mean, X_std, y_mean, y_std)
+
+        self.shift = shift
+        self.scale = scale
+        
+        return self    
+        
+        
+    def predict(self, X):
+        """ Predict performs the z-score bias correction 
+        on the future model dataset, X.
+        
+        Parameters
+        ----------
+        X : DataFrame, shape (n_samples, 1)
+            Samples.
+        window_width : The size of the moving window for 
+            statistical analysis. Default is 31 days.
+        var_str :  The key associated with the target 
+            dataframe variable
+
+        Returns
+        -------
+        fut_corrected : pd.DataFrame, shape (n_samples, 1)
+            Returns corrected values.
+        """
+        fut_mean, fut_std, fut_zscore = _get_fut_stats(X, selfwindow_width)
+        shift_expanded, scale_expanded = _expand_params(S, self.var_str, shift, scale)
+
+        fut_mean_corrected, fut_std_corrected = _correct_fut_stats(fut_mean, fut_std, var_str, shift_expanded, scale_expanded)
+        fut_corrected = (fut_zscore * fut_std_corrected) + fut_mean_corrected
+        return fut_corrected
+  
+
+def _reshape(ds, window_width):
+    """
+    Helper function for `fit` that splits the year and day
+    dimensions of the time-coordinate and bookends the years
+    e.g. (Dec15:31 + whole year + Jan1:15) if window_width is 31 days.
+    ---------------------
+    Parameters:
+    ds : Xarray Dataset
+    window_width: the size of the rolling window.
+
+    Returns:
+    ds_rsh : Reshaped Xarray dataset
+    """
+    split = lambda g: (g.rename({'time': 'day'})
+                       .assign_coords(day=g.time.dt.dayofyear.values))
+    ds_split = ds.groupby('time.year').apply(split)
+    
+    early_Jans = ds_split.isel(day = slice(None,window_width//2))
+    late_Decs = ds_split.isel(day = slice(-window_width//2,None))
+    
+    ds_rsh = xr.concat([late_Decs,ds_split,early_Jans],dim='day')
+    return ds_rsh
+
+
+def _calc_stats(df, window_width):
+    """
+    Helper function for `fit` that calculates the rolling mean and  
+    standard deviation for each day of the year across all years.
+    ---------------------
+    Parameters:
+    df : Pandas dataframe
+    window_width: the size of the rolling window.
+
+    Returns:
+    df_mean : Means for each day of year across all years
+    df_std:  Standard deviations for each day of year across all years
+    """
+    
+    ds = df.to_xarray()
```

Comment:
I think you need to make sure the dataset you create here has a coordinate named time. This probably means renaming the default `index` coordinate name to `time` when necessary.

Reply:
I would want to do that when the data is still in a pandas dataframe format?

> I think you need to make sure the dataset you create here has a coordinate named time. This probably means renaming the default `index` coordinate name to `time` when necessary.

I would want to do that when the data is still in a pandas dataframe format?

Reply:
How do I find out the default name of the index coordinate associated with time?

Reply:
I think you could something like:

```python
if 'time' not in ds.coords and 'index' in ds.coords:
    ds = ds.rename({'index': 'time'})
assert 'time' in ds.coords
```

Reply:
Oh I like that better. I was looking at df.rename, but then I also saw df.set_names.

```
df.set_names('time')
```

Will change it to your method.

Reply:
ahh mine didn't work because I had df.set_names instead of df.Index.set_names. Still I understand your method better and it seems more robust.

Reply:
Okay we're getting closer:

 NameError: name 'selfwindow_width' is not defined

Reply:
@jhamman The error now is similar to with time. I set the name of the variable to 'U' which is not in the tested dataset. So I think we want something similar to:

```
if 'U' not in ds.data_vars ...
```

but I'm working on what that would look like.